File: framework/python/examples/py_record.py
import time
import sys
import os

sys.path.append("../")
from cybertron import pyinit
from cybertron import pyrecord
from proto import record_pb2

#TEST_RECORD_FILE = "p20180903154123.record"
TEST_RECORD_FILE = "test02.record"
CHAN_1 = "channel/chatter"
CHAN_2 = "/test2"
MSG_TYPE = "apollo.cybertron.proto.Test"
STR_10B = "1234567890"
TEST_FILE = "test.record"

def test_record_writer(writer_path):
    fw = pyrecord.RecordWriter()
    if (not fw.Open(writer_path)):
        print("writer open failed!")
        return 
    print("+++ begin to writer...")
    fw.WriteChannel(CHAN_1, MSG_TYPE, STR_10B)
    msg1 = record_pb2.SingleMessage()
    msg1.channel_name = CHAN_1
    msg1.content = STR_10B
    msg1.time = 888
    msg_str = msg1.SerializeToString()
    fw.WriteMessage(msg_str)
    fw.WriteMessage(msg_str)
    fw.Close()

def test_record_reader(reader_path):
    fr = pyrecord.RecordReader()
    if (not fr.Open(reader_path)):
        print("reader open failed!")
        return 
    time.sleep(1)
    print("+"*80)
    print("+++begin to read...")
    count = 0
    read_msg_succ = True
    while not fr.EndOfFile():
        print("="*80)
        print("read [%d] msg" % count)
        read_msg_succ = fr.ReadMessage()
        if(read_msg_succ):
            channelname = fr.CurrentMessageChannelName()
            print("chnanel_name -> %s" % fr.CurrentMessageChannelName())
            #print("msg -> %s" % fr.CurrentRawMessage())
            print("msgtime -> %d" % fr.CurrentMessageTime())
            print("msgnum -> %d" % fr.GetMessageNumber(channelname))
            print("msgtype -> %s" % fr.GetMessageType(channelname))
            #print("pbdesc -> %s" % fr.GetProtoDesc(channelname))
            count = count + 1
    fr.Close()

    pyinit.shutdown()

def main():
    pyinit.init()
    test_record_writer(TEST_RECORD_FILE)
    test_record_reader(TEST_RECORD_FILE)
    pyinit.shutdown()

if __name__ == '__main__':
  main()
  

File: cyber/python/examples/py_record.cc
/******************************************************************************
 * Copyright 2018 The Apollo Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *****************************************************************************/

#include <string>

#include "cyber/cyber.h"
#include "cyber/proto/unit_test.pb.h"
#include "cyber/py_wrapper/py_record.h"

const char TEST_RECORD_FILE[] = "test02.record";
const char CHAN_1[] = "channel/chatter";
const char CHAN_2[] = "/test2";
const char MSG_TYPE[] = "apollo.cyber.proto.Test";
const char STR_10B[] = "1234567890";
const char TEST_FILE[] = "test.record";

void test_write(const std::string &writefile) {
  apollo::cyber::record::PyRecordWriter rec_writer;
  AINFO << "++++ begin writer";
  rec_writer.Open(writefile);
  rec_writer.SetSizeOfFileSegmentation(0);
  rec_writer.SetIntervalOfFileSegmentation(0);
  rec_writer.WriteChannel(CHAN_1, MSG_TYPE, STR_10B);
  rec_writer.WriteMessage(CHAN_1, STR_10B, 1000);
  rec_writer.Close();
}

void test_read(const std::string &readfile) {
  apollo::cyber::record::PyRecordReader rec_reader(readfile);
  AINFO << "++++ begin reading";
  sleep(1);
  int count = 0;

  apollo::cyber::record::BagMessage bag_msg = rec_reader.ReadMessage();
  while (!bag_msg.end) {
    AINFO << "========================";
    std::string channel_name = bag_msg.channel_name;
    AINFO << "read msg[" << count << "]";
    AINFO << "cur channel:[" << channel_name
          << "] msg total:" << rec_reader.GetMessageNumber(channel_name) << "] "
          << "cur msg:[ " << bag_msg.data << " ]";
    AINFO << "curMsgTime: " << bag_msg.timestamp;
    AINFO << "msg type:" << bag_msg.data_type;
    AINFO << "msg protoDesc:" << rec_reader.GetProtoDesc(channel_name);
    count++;
    bag_msg = rec_reader.ReadMessage();
  }

  AINFO << "reader msg count = " << count;
}

// ./py_record readfile1. only read readfile1
// other write & read
int main(int argc, char *argv[]) {
  apollo::cyber::Init("cyber_python");
  if (argc == 2) {
    std::string readfile(argv[1]);
    AINFO << "beging to read: " << readfile;
    test_read(readfile);
    return 1;
  }
  test_write(TEST_RECORD_FILE);
  sleep(1);
  test_read(TEST_RECORD_FILE);

  return 0;
}
